Documentation/design-docs/standalone-gc-loading.md

```diff
@@ -0,0 +1,222 @@
+HRESULT, the GC is considered initialized and the remainder of EE startup continues. If `GC_Initialize` returns
+an error HRESULT, the initialization has failed.
+
+### Initializing the DAC
+
+The existence of a standalone GC is a debuggee process has implications for how the DAC is loaded and
+initializes itself. The DAC has access to implementation details of the GC that are not normally exposed as part
+of the `GC/EE` interfaces, and as such it is versioned differently.
+
+When the DAC is being initialized and it loads the `GcDacVars` structure from the debuggee process's memory, it
+must check the major and minor versions of the DAC, which are itself DAC variables exposed by a standalone GC.
+It then decides whether or not the loaded GC is compatible with the DAC that is currently executing. It does this
+in the same manner that the EE does:
+
+* If the major versions of the DAC and loaded GC do not agree, reject.
+* If the minor version of the DAC is greater than the minor version of the GC, accept but take care
+  not to invoke any new code paths not present in the target GC.
+* If the minor version of the DAC is less than or equal to the minor version of the GC, accept.
+
+If a DAC rejects a loaded GC, it will return `E_FAIL` from DAC APIs that would otherwise need to interact with the
+GC.
+
+### Loader State Machine
+
+The loading of a standalone GC by the EE is governed by a state machine with these states:
+
+* `LOCAL_GC_STATUS_STARTING` - The EE has initiated a standalone GC load, but has not done anything yet.
+* `LOCAL_GC_STATUS_DONE_LOAD` - The EE has looked for a file with `StandaloneGCName` in the directory `StandaloneGCLocation`
+  and called `LoadLibrary` on it successfully.
+* `LOCAL_GC_STATUS_GET_VERSIONINFO` - The EE has successfully located the version information function from the standalone
+  GC.
+* `LOCAL_GC_STATUS_CALL_VERSIONINFO` - The EE has successfully called the version information function and received a 
+  response.
+* `LOCAL_GC_STATUS_VERSION_CHECK` - The EE has performed the version check and it succeeded.
+* `LOCAL_GC_STATUS_GET_INITIALIZER` - The EE has successfully located the GC initialization function.
+* `LOCAL_GC_STATUS_DONE` - The EE has successfully called the GC initialization function and the standalone GC is
+  fully initialized.
+
+If loading a standalone GC fails, the combination of which state the loader was in at the point of failure and
+the HRESULT that indicates *how* that state failed is enough to log a diagnostic indicative of what occured.
+
+## Outstanding Questions
+
+How can we provide the most useful error message when a standalone GC fails to load? In the past it has been difficult
+to determine what preciscely has gone wrong with `coreclr_initialize` returns a HRESULT and no indication of what occured.
+
+Same question for the DAC - Is `E_FAIL` the best we can do? If we could define our own error for DAC/GC version
```

Comment:
I would encourage a more specific HRESULT. Lets use a pragmatic definition of breaking change here - do SOS/ICorDebug tests behave worse with E_SPECIFIC_ERROR than E_FAIL? If so we should re-evaluate but I suspect it will make little difference. My hunch is the bigger issue will be getting tools to fail cleanly when GC makes a breaking change regardless of the HRESULT you use to report the error.

Reply:
sounds good! I'll take a look at this in detail today.

Reply:
SOS would not behave any differently if a failure HRESULT other than `E_FAIL` is returned. I did find this: https://github.com/dotnet/coreclr/blob/master/src/debug/di/process.cpp#L205-L207 - I suppose we would need to add a new error code tot his check or emit `CORDB_E_UNCOMPATIBLE_PLATFORMS` to avoid trouble.

Reply:
Adding a new HRESULT to the di/process.cpp would be my preference. CORDB_UNCOMPATIBLE_PLATFORMS specifically means a difference in hardware architecture and I'd rather not stretch its meaning to include versioning of local GC.

Did you get a chance to see observe the user-visible behavior you get in common tools when do return a failing HRESULT (even if it is E_FAIL)?

Reply:
I've looked at CLRMD and mdbg. CLRMD does not attempt to handle any `COMException` so the buck is passed on to users. Mdbg does handle `COMException`s and, in particular, special-cases E_FAIL in a couple places that I've found to do something other than propagate the error:

* Retrieving the current exception on a managed thread
* Retrieving the current notification on a thread
* Trying to resolve a source location into an IL location
* Trying to find the entry point of an assembly
* Trying to find symbols for a method

None of these would be affected by a non-E_FAIL HRESULT from GC-related DBI methods.

I haven't looked at clrdbg since I'm not sure where the source for that is.

Reply:
clrdbg is a pretty large codebase so I'm not sure how successful you would be at trying to do code review. I'd suggest directly running Visual Studio debugger and seeing what behavior you see when you try to debug an application with a local GC that reports an incompatible version. Assuming that basic breakpoints, stepping, callstacks, locals all work I would probe more on the UI for taking heap snapshots and diffing them. I would also check out what it looks like to use windbg+SOS, in particular the SOS commands that examine the GC heap.

Reply:
Ok! I haven't yet implemented this feature, so would you be OK with keeping this document as-is and then amending it once I have the opportunity to test clrdbg in this manner?

Reply:
I have no qualms about the doc content, feel free to merge at any time.

Reply:
great, thanks!